platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/VirtualTeacherTreeView.java

```diff
@@ -54,11 +54,52 @@ public final class VirtualTeacherTreeView<K extends VirtualKey, V extends Virtua
 
     private static final Logger logger = LogManager.getLogger(VirtualTeacherTreeView.class);
 
+    // A queue of the nodes (by path) that we are about to handle. Note that ConcurrentBitSetQueue
+    // cleans up after itself in "chunks", such that we don't end up consuming a ton of memory.
+    // We use two separate accumulating and processing queues to implement a level-by-level traversal:
+
+    /** A queue to which we add items for the next level currently. */
+    private ConcurrentBitSetQueue accumulatingHandleQueue = new ConcurrentBitSetQueue();
+
+    /** A queue which we're processing currently (aka sending nodes to the learner). */
+    private ConcurrentBitSetQueue processingHandleQueue = new ConcurrentBitSetQueue();
+
+    /** Flip the queues by exchanging the accumulating and processing queues. */
+    private synchronized void flipQueues() {
+        final ConcurrentBitSetQueue temp = accumulatingHandleQueue;
+        accumulatingHandleQueue = processingHandleQueue;
+        processingHandleQueue = temp;
+    }
+
+    /**
+     * A node which status has to be reported by the learner before the teacher resumes sending
+     * the next level of the tree.
+     * The node can end up being either KNOWN or NOT_KNOWN, either explicitly or implicitly
+     * by inferring the status from the previously reported ancestors' statuses (except
+     * the root status because it's always implicitly NOT_KNOWN, so we ignore it.)
+     * <p>
+     * It's volatile to avoid extra locks when accessing the value. It's totally
+     * okay if the value is updated after it's read. The worst that could happen
+     * is that we'll start sending the next level of nodes before receiving
+     * a response for the last node at the current level, but this is unlikely to happen
+     * because the code processing the queue is single-threaded,
+     * and it doesn't hurt from the correctness perspective even if it happens.
+     */
+    private volatile Long lastNodeAwaitingReporting = null;
+
+    /**
+     * A node status may become implicitly reported by a response about its ancestors,
+     * in which case we can retest the condition and go ahead and send the next level
+     * without waiting for this particular node to report its status since it's already known/inferred.
+     * So we use a timeout to recheck the condition periodically.
+     */
+    private static final long AWAIT_FOR_REPORT_TIMEOUT_MILLIS = 100;
```

Comment:
100ms is too coarse-grained. Since `wait()` is called in a loop, and total wait time is checked against MAX_TOTAL_... anyway, 1ms can be used here.

Reply:
Done, and also decreased the MAX_ timeout to 1 second.